DOCSP-46425: Specify a Query (#2)
mongodb/docs-django: source/interact-data/specify-a-query.txt

.. _django-specify-query:

===============
Specify a Query
===============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: expressions, operations, read, filter, code example

Overview
--------

In this guide, you can learn how to use {+django-odm+} to specify
a database query.

> The Django ``QuerySet`` API provides functions that allow you to retrieve

Reviewer: s/functions/methods/

> model objects. To run a MongoDB database query, call ``QuerySet`` functions

Reviewer: s/functions/methods/

on your model's ``Manager``. The ``Manager`` class handles database
operations and allows you to interact with your MongoDB data by referencing
Django models. By default, Django adds a ``Manager`` named ``objects``
to every model class. 

When you assign a ``QuerySet`` to a variable, {+django-odm+} does not

> perform the operation until you evaluate the variable, usually by printing. After

Reviewer: s/,usually by printing//.

I would leave this out, more here about how you can evaluate: https://docs.djangoproject.com/en/5.1/ref/models/querysets/#when-querysets-are-evaluated

complex matching criteria.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``Movie`` model, which represents
the ``sample_mflix.movies`` collection from the :atlas:`Atlas sample datasets </sample-data>`.
The ``Movie`` model contains an embedded ``Award`` model as the value
of its ``awards`` field. These model classes have the following definitions:

.. literalinclude:: /includes/interact-data/specify-a-query.py
   :start-after: start-models
   :end-before: end-models
   :language: python
   :copyable:

You can use the Python interactive shell to run the code examples.
To enter the shell, run the following command from your project's 
root directory:

.. code-block:: bash

   python manage.py shell

After entering the Python shell, ensure that you import the following models and
modules:

.. code-block:: python

   from <your application name>.models import Movie, Award
   from django.utils import timezone
   from datetime import datetime

To learn how to create a Django application that uses the ``Movie``
model and the Python interactive shell to interact with MongoDB documents,
visit the :ref:`django-get-started` tutorial.

Run a Query
-----------

> To query your MongoDB data, call a ``QuerySet`` function on your 

Reviewer: s/function/method/

model's ``Manager`` class and specify your matching criteria in a
query filter.

> This section describes how to use the following functions from

Reviewer: s/functions/methods/

the ``QuerySet`` API:

- :ref:`all() <django-query-retrieve-all>`
- :ref:`filter() <django-query-retrieve-matching>`
- :ref:`get() <django-query-retrieve-one>`
- :ref:`exclude() <django-query-exclude>`

.. _django-query-retrieve-all:

Retrieve All Documents
~~~~~~~~~~~~~~~~~~~~~~

To retrieve all documents from a collection, call the ``all()``

> function on your model's ``Manager`` class.

Reviewer: s/function/method/

> The following example calls the ``all()`` function to retrieve

Reviewer: s/function/method/

all documents in the ``sample_mflix.movies`` collection:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-all
      :end-before: end-all
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: The Great Train Robbery>, <Movie: A Corner in Wheat>,
      <Movie: Winsor McCay, the Famous Cartoonist of the N.Y. Herald and His Moving Comics>,
      <Movie: Traffic in Souls>, <Movie: Gertie the Dinosaur>,
      <Movie: In the Land of the Head Hunters>, <Movie: The Perils of Pauline>,
      <Movie: The Italian>, <Movie: Regeneration>, <Movie: Civilization>,
      <Movie: Where Are My Children?>, <Movie: The Poor Little Rich Girl>,
      <Movie: Wild and Woolly>, <Movie: The Blue Bird>, <Movie: From Hand to Mouth>,
      <Movie: High and Dizzy>, <Movie: One Week>, <Movie: The Saphead>,
      <Movie: The Ace of Hearts>, <Movie: The Four Horsemen of the Apocalypse>,
      '...(remaining elements truncated)...']>

.. _django-query-retrieve-matching:

Retrieve Matching Documents
~~~~~~~~~~~~~~~~~~~~~~~~~~~

To query a collection for documents that match a set of criteria,

> call the ``filter()`` function on your model's ``Manager`` class.

Reviewer: s/function/method/

> Pass a query filter to the ``filter()`` function that specifies your

Reviewer: s/function/method/

query criteria.

> The following example calls the ``filter()`` function

Reviewer: s/function/method/

to query the ``sample_mflix.movies`` collection
for documents that have a ``runtime`` value of ``300``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-filter
      :end-before: end-filter
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: Wild Palms>, <Movie: Streets of Laredo>,
      <Movie: The Way We Live Now>]>

.. _django-query-retrieve-one:

Retrieve One Document
~~~~~~~~~~~~~~~~~~~~~

To retrieve one document from a collection, call the ``get()``

> function on your model's ``Manager`` class. Pass 

Reviewer: s/function/method/

> a query filter to the ``get()`` function that specifies your

Reviewer: s/function/method/

query criteria.

.. important::

   If your query matches no documents or multiple documents, the ``get()``

>    function generates an error. To retrieve one document from a query

Reviewer: s/function/method/

   that might match multiple, use the :ref:`first() <django-query-first>`

>    function.

Reviewer: s/function/method/

> The following example calls the ``get()`` function

Reviewer: s/function/method/

to retrieve one document that has a ``title``
value of ``"Finding Nemo"``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-get
      :end-before: end-get
      :language: python

   .. output::
      :visible: false

      <Movie: Finding Nemo>

.. _django-query-exclude:

Exclude Matching Documents
~~~~~~~~~~~~~~~~~~~~~~~~~~

To query a collection for documents that do not meet your

> search criteria, call the ``exclude()`` function on your model's

Reviewer: s/function/method/

``Manager`` class. Pass the exclusion criteria as an 

> argument to the ``exclude()`` function.

Reviewer: s/function/method/

> The following example calls the ``exclude()`` function to

Reviewer: s/function/method/

field you want to query, as shown in the following code:

.. code-block:: python
   :copyable: false

   Model.objects.filter(<model field>__<related model field>=<value>)

Example
```````

This example uses a lookup that spans the ``Movie`` and ``Award``
relationship. The ``Movie`` model's ``awards`` field has an
embedded ``Award`` model value. This embedded model represents
the ``awards``  object field in the ``sample_mflix.movies`` collection
and its nested ``wins``,  ``nominations``, and ``text`` fields. The
following code uses a lookup to query for documents in which the ``awards.wins``
nested field value is ``93``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-filter-relationships
      :end-before: end-filter-relationships
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: The Queen>, <Movie: Dallas Buyers Club>]>

.. _django-query-combine:

Combine Lookups
~~~~~~~~~~~~~~~

You can combine field lookups to further specify your
query criteria. To combine lookups, chain each lookup
type together and separate them with a double underscore

> (``__``).

Author: Alright I made some changes - described Q objects in this section instead as a way to combine query filters, and moved the embedded model field section to the "Advanced Field Queries" section

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-filter-combine
      :end-before: end-filter-combine
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: Hallelujah>, <Movie: Little Caesar>,
      <Movie: Morocco>, <Movie: Romance>, <Movie: The Guardsman>,
      <Movie: The Front Page>, <Movie: The Public Enemy>,
      <Movie: è Nous la Libertè>, <Movie: Smilin' Through>,
      <Movie: Flying Down to Rio>, <Movie: She Done Him Wrong>,
      <Movie: State Fair>, <Movie: The Barretts of Wimpole Street>,
      <Movie: Imitation of Life>, <Movie: The Lost Patrol>,
      <Movie: Alice Adams>, <Movie: Black Fury>, <Movie: David Copperfield>,
      <Movie: Les Misèrables>, <Movie: Roberta>,
      '...(remaining elements truncated)...']>

Modify Query Results
--------------------

This section describes how to modify your query
results in the following ways:

- :ref:`django-query-sort`
- :ref:`django-query-limit`
- :ref:`django-query-first`

.. _django-query-sort:

Sort Results
~~~~~~~~~~~~

You can provide a sort order for your query results by using the

> ``order_by()`` function. To specify an ascending sort based on values

Reviewer: s/function/method/

   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-sort
      :end-before: end-sort
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: Rocky>, <Movie: Rocky II>, <Movie: Rocky III>,
      <Movie: Rocky IV>, <Movie: Rocky V>, <Movie: Rocky Marciano>,
      <Movie: Rocky Balboa>]>

.. _django-query-limit:

Limit Results
~~~~~~~~~~~~~

You can specify the number of results that a query returns
by using Python's array-slicing syntax, as shown
in the following code:

.. code-block:: python
   :copyable: false

   Model.objects.filter(<query filter>)[<start>:<end>]

Replace the ``<start>`` and ``<end>`` placeholders with integer
values representing the subset of results you want to return.
The start value is non-inclusive and the end value is inclusive.

If you omit the ``<start>`` value, the query returns each result,
beginning with the first match, until it returns the number specified
by the ``<end>`` value.

If you omit the ``<end>`` value, the query returns each result
but skips the number of initial results specified by ``<start>``.

Example
```````

This example performs the following actions:

> - Calls the ``filter()`` function on the ``Movie`` model's ``Manager`` to query

Reviewer: s/function/method/

  the ``sample_mflix.movies`` collection
- Queries documents that have a ``released`` value of 
  ``datetime(2010, 7, 16)`` (July 16, 2010)
- Returns the third and fourth results

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-limit
      :end-before: end-limit
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: Inception>, <Movie: Winter's Bone>]>

.. _django-query-first:

Retrieve the First Result
~~~~~~~~~~~~~~~~~~~~~~~~~

To retrieve the first result from a query that might match

> multiple results, chain the ``first()`` function to the ``filter()``

Reviewer: s/function/method/

> function. Pass a query filter to the ``filter()`` function that specifies your

Reviewer: s/function/method/

query criteria.

> The following example calls the ``filter()`` and ``first()`` functions

Reviewer: s/functions/methods/

to query the ``sample_mflix.movies`` collection for the first
document in which the ``genres`` value is  ``["Crime", "Comedy"]``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-first
      :end-before: end-first
      :language: python

   .. output::
      :visible: false

      <Movie: The Crew>

Advanced Field Queries
----------------------

This section describes how to run queries on the

> following field types:

Author: Added!

to query.

The following example queries the ``imdb`` field, a ``JSONField``,
for values of its nested ``votes`` field that exceed ``900000``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-json
      :end-before: end-json
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: La nao capitana>, <Movie: The Godfather>,
      <Movie: This Is Spinal Tap>, <Movie: Forrest Gump>, <Movie: Pulp Fiction>,
      <Movie: The Shawshank Redemption>, <Movie: Se7en>,
      <Movie: The Lord of the Rings: The Fellowship of the Ring>,
      <Movie: The Matrix>, <Movie: Fight Club>,
      <Movie: The Lord of the Rings: The Return of the King>,
      <Movie: The Lord of the Rings: The Two Towers>, <Movie: The Shawshank Redemption>,
      <Movie: Landet som icke èr>, <Movie: The Dark Knight>,
      <Movie: The Danish Girl>, <Movie: The Dark Knight Rises>, <Movie: Inception>,
      <Movie: Catching the Sun>, <Movie: Scouts Guide to the Zombie Apocalypse>,
      '...(remaining elements truncated)...']>

JSONField Transforms
````````````````````

> You can use the ``KT()`` function to specify a key, index, or path

Reviewer: For reasons I don't fully understand, `KT` is an expression not a function: https://docs.djangoproject.com/en/5.1/topics/db/queries/#module-django.db.models.fields.json

transform of a ``JSONField``. For example, the following code uses
the ``annotate()`` and ``KT()`` functions to create a new ``score`` key, which
stores ``imdb.rating`` nested field values. Then, the code sorts 
each document in the ``sample_mflix.movies`` collection by 

> their descending ``score`` values:

Reviewer 2: Adding two tips:
* One for a short explanation of KT -- used to work with the text value of keys, indexes, or path transforms within a JSONField. With a link to the KT documentation in Django
* One for [annotate](https://docs.djangoproject.com/en/5.1/ref/models/querysets/#django.db.models.query.QuerySet.annotate) (very powerful method!) -- Annotates each object in the QuerySet with the provided list of [query expressions](https://docs.djangoproject.com/en/5.1/ref/models/expressions/).

You can use the ``pk`` lookup shortcut to query primary key
values, which MongoDB stores as ``ObjectId`` values.

The following example queries the ``sample_mflix.movies`` collection
for a document whose primary key is ``ObjectId("573a1394f29313caabce0d37")``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-primary-key-pk
      :end-before: end-primary-key-pk
      :language: python

   .. output::
      :visible: false
     
      // Your ObjectId values might differ

      <Movie: Vertigo>

The ``Movie`` model, which represents the ``sample_mflix.movies`` collection,
uses the ``id`` field as its primary key. The following example constructs
the same query as the preceding code:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-primary-key-id
      :end-before: end-primary-key-id
      :language: python

   .. output::
      :visible: false
     
      // Your ObjectId values might differ

      <Movie: Vertigo>

Additional Information
----------------------

To learn how to run raw database queries using MongoDB's aggregation 

> pipeline syntax, rather than ``QuerySet`` API functions, see the

Reviewer: s/functions/methods/

         value.

Example
```````

The following example uses the ``__lte`` lookup to 
query for documents in which the ``runtime`` value
is less than or equal to ``50``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-filter-lte
      :end-before: end-filter-lte
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: The Great Train Robbery>, <Movie: A Corner in Wheat>,
      <Movie: Winsor McCay, the Famous Cartoonist of the N.Y. Herald and His Moving Comics>,
      <Movie: Gertie the Dinosaur>, <Movie: From Hand to Mouth>,
      <Movie: High and Dizzy>, <Movie: One Week>, <Movie: Now or Never>,
      <Movie: Steamboat Willie>, <Movie: The Music Box>, <Movie: Three Little Pigs>,
      <Movie: The Band Concert>, <Movie: Who Killed Cock Robin?>,
      <Movie: Popeye the Sailor Meets Sindbad the Sailor>, <Movie: Dots>, <Movie: The Land>,
      <Movie: Meshes of the Afternoon>, <Movie: The Memphis Belle: A Story of a Flying Fortress>,
      <Movie: Report from the Aleutians>, <Movie: Saludos Amigos>,
      '...(remaining elements truncated)...']>

.. _django-query-combine:

> Combine Lookups

Reviewer 2: This is beautiful documentation on using the `Q` Objects. I wanted to also add there are three ways common ways to combine lookups:

Using [multiple kwargs](https://docs.djangoproject.com/en/5.1/topics/db/queries/#retrieving-objects)
```python
Model.objects.filter(<field>_<lookup_type>=<value>, <field>_<lookup_type>=<value>, ...)
```
Using [Q Objects](https://docs.djangoproject.com/en/5.1/topics/db/queries/#complex-lookups-with-q-objects)
```python
Model.objects.filter(
Q(<field>__<lookup type>=<value>) <operator> Q(<field>__<lookup type>=<value>)
)
```
[Filter chaining](https://docs.djangoproject.com/en/5.1/topics/db/queries/#chaining-filters)
```python
Model.objects.filter(<field>_<lookup_type>=<value>). filter(<field>_<lookup_type>=<value>, ...)
```

Chaining is also cool because you can chain _any_ QuerySet method with another. I.e. 
```python
all = Model.objects.all()
most = all.filter()
few = most.exclude()
fewer = few.filter()
```
All three are important to raise. For them I would give each a blurb under this `Combine Lookups` section. 

Again, great work. :)

Author: I listed all three in this section, but linked out for additional information rather than providing my own examples (so the scope of this page doesn't get too big)

      '...(remaining elements truncated)...']>

.. _django-query-primary-key:

Query a Primary Key Field
~~~~~~~~~~~~~~~~~~~~~~~~~

You can use the ``pk`` lookup shortcut to query primary key
values, which MongoDB stores as ``ObjectId`` values.

The following example queries the ``sample_mflix.movies`` collection
for a document whose primary key is ``ObjectId("573a1394f29313caabce0d37")``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-primary-key-pk
      :end-before: end-primary-key-pk
      :language: python

   .. output::
      :visible: false
     
      // Your ObjectId values might differ

      <Movie: Vertigo>

The ``Movie`` model, which represents the ``sample_mflix.movies`` collection,
uses the ``id`` field as its primary key. The following example constructs
the same query as the preceding code:

.. literalinclude:: /includes/interact-data/specify-a-query.py
    :language: python
    :dedent:
    :start-after: start-primary-key-id
    :end-before: end-primary-key-id

Additional Information
----------------------

To learn how to run raw database queries by using MongoDB's aggregation 

> pipeline syntax, see the :ref:`django-raw-queries` guide.

Reviewer 2: Two things:
1. I think this blurb should get moved to the "Advanced Field Queries" section. 
2. We should still point to the `django-raw-queries` guide, but I think we should adopt something more like the language in [Django's documentation](https://docs.djangoproject.com/en/5.1/topics/db/queries/#falling-back-to-raw-sql).

The blurb could go something like this...
```txt
To leverage the power of querying in Django using the MongoDB Aggregation Pipeline directly, 
you can fall back on doing saw through our raw querying.
Check out our section on raw querying; see the :ref:`django-raw-queries` guide.
```

      <Movie: In the Land of the Head Hunters>, <Movie: The Perils of Pauline>,
      <Movie: The Italian>, <Movie: Regeneration>, <Movie: Civilization>,
      <Movie: Where Are My Children?>, <Movie: The Poor Little Rich Girl>,
      <Movie: Wild and Woolly>, <Movie: The Blue Bird>, <Movie: From Hand to Mouth>,
      <Movie: High and Dizzy>, <Movie: One Week>, <Movie: The Saphead>,
      <Movie: The Ace of Hearts>, <Movie: The Four Horsemen of the Apocalypse>,
      '...(remaining elements truncated)...']>

.. _django-query-retrieve-matching:

Retrieve Matching Documents
~~~~~~~~~~~~~~~~~~~~~~~~~~~

To query a collection for documents that match a set of criteria,
call the ``filter()`` method on your model's ``Manager`` class.
Pass a query filter to the ``filter()`` method that specifies your
query criteria.

The following example calls the ``filter()`` method
to query the ``sample_mflix.movies`` collection
for documents that have a ``runtime`` value of ``300``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-filter
      :end-before: end-filter
      :language: python

   .. output::
      :visible: false

>       <QuerySet [<Movie: Wild Palms>, <Movie: Streets of Laredo>,

Author: The Movie model's __str__() method (in the sample data section) defines its string representation; since that method returns the title field, you only see the title anytime you output results

.. _django-query-raw:

Run Raw Database Queries
~~~~~~~~~~~~~~~~~~~~~~~~

If you want to run complex queries that Django's query API
cannot express, you can use the ``raw_aggregate()`` method. This
method allows you to specify your query criteria in a MongoDB
aggregation pipeline, which you pass as an argument to 
``raw_aggregate()``. 

To learn how to run raw database queries, see the :ref:`django-raw-queries`
guide.

Customize Your Query Filter
---------------------------

You can further refine your query criteria by using a **field lookup**
in your query filter. Field lookups allow you to clarify the relationship
between the field you're querying and the value you want to
query for. Use the following syntax to specify a field lookup:

.. code-block:: python
   :copyable: false

   Model.objects.filter(<field name>__<lookup type>=<value>)
   
You can use different lookup types to perform advanced queries,
such as partial text matching, array element queries, regular
expression matching, and year value matching for datetime fields. 

.. tip::

   To view a full list of lookup types, see `Field lookups
   <{+django-docs+}/ref/models/querysets/#field-lookups>`__ in the

>    ``QuerySet`` Django documentation.

Reviewer 3: ```suggestion
   ``QuerySet`` Django API reference.
```

This section describes how to refine your query filters
in the following ways:

- :ref:`django-query-text-match`
- :ref:`django-query-comparison`
- :ref:`django-query-combine`

.. _django-query-text-match:

Use Text Match Lookups
~~~~~~~~~~~~~~~~~~~~~~

The following table describes some of the lookup types
that you can use to run queries that include specific
matching criteria for text values:

.. list-table::
   :header-rows: 1
   :widths: 20 80

   * - Lookup Type
     - Description

>    * - ``__exact``

Reviewer 3: Q: would it be better to exclude the underscores from these list items, as this is how they appear in the API reference?

Author: yes good point, I think since the Django docs doesn't include them I will remove

that meet the following query conditions:

- ``title`` field value begins either with the text ``"Funny"``
  or ``"Laugh"``
-  ``genres`` array field does not contain the value ``"Comedy"``

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-filter-combine
      :end-before: end-filter-combine
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: Laugh, Clown, Laugh>, <Movie: Funny Games>]>

.. tip:: 

   The preceding example uses a field lookup on an
   ``ArrayField`` value. For more information about querying
   an ``ArrayField``, see the :ref:`django-query-arrayfield`
   section in this guide.

Modify Query Results
--------------------

This section describes how to modify your query
results in the following ways:

- :ref:`django-query-sort`
- :ref:`django-query-limit`
- :ref:`django-query-first`

.. _django-query-sort:

Sort Results
~~~~~~~~~~~~

You can provide a sort order for your query results by using the
``order_by()`` method. To specify an ascending sort based on values
of a given field, pass the field name as an argument. To specify a descending

> sort, pass the field name prefixed with a minus symbol (``-``) as an argument.

Reviewer 3: S: mention if you can chain sorts

Author: yes you can pass multiple fields; added

Example
```````

This example performs the following actions:

- Calls the ``filter()`` method on the ``Movie`` model's ``Manager`` to query
  the ``sample_mflix.movies`` collection
- Queries documents that have a ``title`` value starting
  with the text ``"Rocky"``
- Sorts the results in ascending order of their ``released`` field
  values

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-sort
      :end-before: end-sort
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: Rocky>, <Movie: Rocky II>, <Movie: Rocky III>,
      <Movie: Rocky IV>, <Movie: Rocky V>, <Movie: Rocky Marciano>,
      <Movie: Rocky Balboa>]>

.. _django-query-limit:

Limit Results

> ~~~~~~~~~~~~~

Reviewer 3: S: change to Skip and Limit Results

You can specify the number of results that a query returns

> by using Python's array-slicing syntax, as shown

Reviewer 3: S: up to you, if its obvious, but can mention that it uses 0-indexing

in the following code:

.. code-block:: python
   :copyable: false

   Model.objects.filter(<query filter>)[<start>:<end>]

Replace the ``<start>`` and ``<end>`` placeholders with integer
values representing the subset of results you want to return.
The start value is non-inclusive and the end value is inclusive.

If you omit the ``<start>`` value, the query returns each result,
beginning with the first match, until it returns the number specified
by the ``<end>`` value.

If you omit the ``<end>`` value, the query returns each result
but skips the number of initial results specified by ``<start>``.

Example
```````

This example performs the following actions:

- Calls the ``filter()`` method on the ``Movie`` model's ``Manager`` to query
  the ``sample_mflix.movies`` collection
- Queries documents that have a ``released`` value of 

>   ``datetime(2010, 7, 16)`` (July 16, 2010)

Reviewer 3: S: in a preceding instance, you didnt provide both the datetime instance and the interpretation, consider standardizing this in this page

Author: I'll provide both - will standardize that across other pages

method. Pass a query filter to the ``filter()`` method that specifies your
query criteria.

The following example calls the ``filter()`` and ``first()`` methods
to query the ``sample_mflix.movies`` collection for the first
document in which the ``genres`` value is  ``["Crime", "Comedy"]``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-first
      :end-before: end-first
      :language: python

   .. output::
      :visible: false

      <Movie: The Crew>

Advanced Field Queries
----------------------

This section describes how to run queries on the
following field types:

- :ref:`EmbeddedModelField <django-query-embedded>`
- :ref:`ArrayField <django-query-arrayfield>`
- :ref:`JSONField <django-query-jsonfield>`
- :ref:`Primary Key <django-query-primary-key>`

.. _django-query-embedded:

Query an EmbeddedModelField
~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can run operations on your model that query fields of embedded
models. To query these fields, specify your model field and each embedded field
name separated by double underscores (``__``) until you reach the 

> field you want to query, as shown in the following code:

Reviewer 3: S: mention if you can use lookup methods as usual as the example doesnt use one (or you can make an example that includes both?)

Author: Yes you can use them, updated

.. code-block:: python
   :copyable: false

   Model.objects.filter(<model field>__<embedded model field>=<value>)

Example
```````

This example uses a lookup that spans the ``Movie`` and ``Award``
relationship. The ``Movie`` model's ``awards`` field has an
embedded ``Award`` model value. This embedded model represents
the ``awards``  object field in the ``sample_mflix.movies`` collection

> and its nested ``wins``,  ``nominations``, and ``text`` fields. The

Reviewer 3: S: newline before example intro

following code uses a lookup to query for documents in which the ``awards.wins``
nested field value is ``93``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-filter-relationships
      :end-before: end-filter-relationships
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: The Queen>, <Movie: Dallas Buyers Club>]>

.. _django-query-arrayfield:

Query an ArrayField

> ~~~~~~~~~~~~~~~~~

Reviewer 3: S: extend underline

You can query data stored in an ``ArrayField`` value
by using the standard query syntax. {+django-odm+} provides

> additional field lookup types for querying ``ArrayField`` values,
> as described in the following table:

Reviewer 3: ```suggestion
additional field lookup types for querying ``ArrayField`` values,
which are described in the following table:
```

.. list-table::
   :header-rows: 1
   :widths: 20 80

   * - Lookup Type
     - Description

   * - ``__contains``
     - | Matches fields that store the provided value as
         a subset of their data. This ``ArrayField`` lookup overrides the
         ``__contains`` lookup described in the :ref:`django-query-text-match`
         section of this guide.

   * - ``__contained_by``
     - | Matches fields that store a subset of the provided values.
         This lookup type is the inverse of ``__contains``.
       
   * - ``__overlap``
     - | Matches field that store any of the provided values.

   * - ``__len``
     - | Matches fields that store the number of values provided.

Example
```````

The following example uses the ``__overlap`` lookup to query

> for documents whose  ``genres`` field stores any values in the

Reviewer 3: ```suggestion
for documents whose  ``genres`` field contains any values in the
```

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-array
      :end-before: end-array
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: The Poor Little Rich Girl>, <Movie: Robin Hood>,
      <Movie: Peter Pan>, <Movie: The Thief of Bagdad>, <Movie: Clash of the Wolves>,
      <Movie: Beau Geste>, <Movie: The Black Pirate>, <Movie: The Son of the Sheik>,
      <Movie: Steamboat Willie>, <Movie: The Big Trail>, <Movie: The Champ>,
      <Movie: Skippy>, <Movie: Tabu: A Story of the South Seas>, <Movie: Movie Crazy>,
      <Movie: Shanghai Express>, <Movie: Tarzan the Ape Man>, <Movie: The Son of Kong>,
      <Movie: Three Little Pigs>, <Movie: Wild Boys of the Road>, <Movie: The Black Cat>,
      '...(remaining elements truncated)...']>

.. _django-query-jsonfield:

Query a JSONField
~~~~~~~~~~~~~~~~~

You can query data stored in a ``JSONField`` value by using the 
same syntax as show in the :ref:`django-query-embedded`
section. Chain each field and nested field name together, separated
by double underscores (``__``), until you reach the field you want
to query.

> The following example queries the ``imdb`` field, a ``JSONField``,
> for values of its nested ``votes`` field that exceed ``900000``:

Reviewer 3: ```suggestion
The following example queries the ``JSONField``-valued ``imdb`` field
for values of its nested ``votes`` field that exceed ``900000``:
```

      <Movie: Coming to Terms>, <Movie: La nao capitana>, <Movie: All Eyes and Ears>,
      <Movie: Catching the Sun>, <Movie: Manhattan Romance>, <Movie: Anomalisa>,
      <Movie: Outliving Emily>, <Movie: Mary Loss of Soul>,
      <Movie: The Childhood of a Leader>, <Movie: Krot na more>,
      '...(remaining elements truncated)...']>

.. _django-query-primary-key:

Query a Primary Key Field
~~~~~~~~~~~~~~~~~~~~~~~~~

You can use the ``pk`` lookup shortcut to query primary key
values, which MongoDB stores as ``ObjectId`` values.

The following example queries the ``sample_mflix.movies`` collection
for a document whose primary key is ``ObjectId("573a1394f29313caabce0d37")``:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-primary-key-pk
      :end-before: end-primary-key-pk
      :language: python

   .. output::
      :visible: false
     
      // Your ObjectId values might differ

      <Movie: Vertigo>

The ``Movie`` model, which represents the ``sample_mflix.movies`` collection,
uses the ``id`` field as its primary key. The following example constructs

> the same query as the preceding code:

Reviewer 3: ```suggestion
the same query as the preceding code by using the ``id`` field:
```

on your model's manager. The ``Manager`` class handles database
operations and allows you to interact with your MongoDB data by referencing
{+framework+} models. By default, {+framework+} adds a ``Manager`` named ``objects``
to every model class. 

When you assign a ``QuerySet`` to a variable, {+django-odm+} does not
perform the operation until you evaluate the variable. After
evaluating the ``QuerySet``, {+framework+} saves the query results in the ``QuerySet``
cache. Future evaluations of the ``QuerySet`` use the cached results.

.. tip::

   To learn more about the {+framework+} ``QuerySet`` API, see `QuerySet API reference
   <{+django-docs+}/ref/models/querysets/>`__ in the {+framework+} documentation.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``Movie`` model, which represents
the ``sample_mflix.movies`` collection from the :atlas:`Atlas sample datasets </sample-data>`.
The ``Movie`` model contains an embedded ``Award`` model as the value
of its ``awards`` field. These model classes have the following definitions:

.. literalinclude:: /includes/interact-data/specify-a-query.py
   :start-after: start-models
   :end-before: end-models
   :language: python

>    :copyable:

Reviewer 3: S: cant comment on the code example but it might be helpful to note either as a code comment or somewhere in this text that the str method means that the string representation of a model is just its title

Retrieve All Documents
~~~~~~~~~~~~~~~~~~~~~~

To retrieve all documents from a collection, call the ``all()``
method on your model's manager.

The following example calls the ``all()`` method to retrieve
all documents in the ``sample_mflix.movies`` collection:

.. io-code-block::
   :copyable:

   .. input:: /includes/interact-data/specify-a-query.py
      :start-after: start-all
      :end-before: end-all
      :language: python

   .. output::
      :visible: false

      <QuerySet [<Movie: The Great Train Robbery>, <Movie: A Corner in Wheat>,
      <Movie: Winsor McCay, the Famous Cartoonist of the N.Y. Herald and His Moving Comics>,
      <Movie: Traffic in Souls>, <Movie: Gertie the Dinosaur>,
      <Movie: In the Land of the Head Hunters>, <Movie: The Perils of Pauline>,
      <Movie: The Italian>, <Movie: Regeneration>, <Movie: Civilization>,
      <Movie: Where Are My Children?>, <Movie: The Poor Little Rich Girl>,
      <Movie: Wild and Woolly>, <Movie: The Blue Bird>, <Movie: From Hand to Mouth>,
      <Movie: High and Dizzy>, <Movie: One Week>, <Movie: The Saphead>,
      <Movie: The Ace of Hearts>, <Movie: The Four Horsemen of the Apocalypse>,

>       '...(remaining elements truncated)...']>

Reviewer 3: S: applies to multiple examples, but up to you to truncate this further as the number of results doesn't signify anything
